dev-docs/RFCs/v7.x/layer-shader-hook-rfc.md

```diff
@@ -0,0 +1,107 @@
+
+- The subclass is difficult to read, as it does not contain the full shader code.
+- The injection usually depends on certain assumptions about the base layer implementation, e.g. attribute names and varyings, which is prone to breakage in a future release.
+- Because each layer's attributes are different, the injection code cannot be ported to other layers easily.
+- The standard injection points (`decl`, `main-start`, `main-end`) are very limiting. If the subclass wishes to manipulate intermediate values in the shader, e.g. `offset` in `ScatterplotLayer`'s vertex shader before projection is performed, custom code must be injected using key strings in the base shader, which is even more fragile.
+
+## Target Use Cases
+
+Publicly documented shader hooks make it possible to:
+
+- Standardize layer shaders and make them easier to tweak for people who are less familiar with WebGL.
+- Write one set of injections that works for many layers.
+- Write custom layers that remain working for future releases of the same major version.
+- As a result of the above, package and distribute layers and layer extensions as reusable components.
+
+### Example 1: Generic GPU Data Filter
+
+The basic implementation of the [GPU data filter](/dev-docs/RFCs/v6.0/data-filter-rfc.md) discards any fragment from objects that are filtered out. However, some users (including kepler.gl) wish to have more control of the output appearance, e.g. circles grow/shrink as they enter/exit a time window; render the objects that are filtered out with a different color; etc. This requires layers to expose shader hooks for object dimensions and colors.
+
+### Example 2: Generic Brushing
+
+Brushing is based on mouse position and the position of the current object. To implement it in a way that is reusable across multiple layers, it requires the layers to expose shader hooks for object positions.
+
+### Example 3: Pattern Fill
+
+Instead of filling an area (triangle) with a solid color, fill it with a tiling pattern. The pattern would be supplied with a sprite image, with additional attributes controlling the size and the orientation. This is needed in rendering realistic maps e.g. a polygon or path that represents "crosswalk." This requires layers to expose shader hooks for object colors and texture coordinates.
+
+## Proposal
+
+### Naming Conventions
+
+Shader hook names follow the pattern `DECKGL_<action>_<target>`.
+
+- All hooks start with the `DECKGL_` prefix
+- Actions:
+  + `FILTER`: if the hook can modify the input value(s).
+  + `USE`: if the hook does not expect any output.
+- Targets:
+  + Colors are normalized to the WebGL color space.
+  + Coordinates are presumed to be in common space.
+  + Non-commonspace coordinates must contain modifiers in their names, e.g. `worldPosition`, `pixelSize`.
+
+### Standard Layer Hooks
+
+#### vs:DECKGL_FILTER_SIZE(inout vec3 size)
```

Comment:
how about DECKGL_FILTER_COMMON_SPACE_DIMENSION? DECKGL_FILTER_SIZE is pretty vague.

Reply:
All shader names default to common space if not specified, see the "naming conventions" section.